From 4aa9eef100e1dcbe119a704b542687bab882fcdb Mon Sep 17 00:00:00 2001
From: Andrzej Surdej <Andrzej_Surdej@comcast.com>
Date: Wed, 23 Sep 2026 14:24:01 +0200
Subject: [PATCH] [GStreamer][MSE] Flush track queues on MediaSource reopen

Reopen MediaSource using a GStreamer pipeline flush to clear the EOS state.
Print a warning if the reopening occurs during active playback.
---
 .../mse/MediaPlayerPrivateGStreamerMSE.cpp         |  3 +++
 .../gstreamer/mse/MediaSourcePrivateGStreamer.cpp  | 14 ++++++++++++++
 .../gstreamer/mse/WebKitMediaSourceGStreamer.cpp   |  5 ++---
 3 files changed, 19 insertions(+), 3 deletions(-)

diff --git a/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp b/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp
index 8e66b7f55f1a..caabfea7cd19 100644
--- a/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp
+++ b/Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp
@@ -563,6 +563,9 @@ void MediaPlayerPrivateGStreamerMSE::setEosWithNoBuffers(bool eosWithNoBuffers)
             // error count integrity anymore.
         } else
             m_ignoreErrors = false;
+    } else {
+        GST_DEBUG_OBJECT(pipeline(), "Clear m_isEndReached.");
+        m_isEndReached = false;
     }
 }
 
diff --git a/Source/WebCore/platform/graphics/gstreamer/mse/MediaSourcePrivateGStreamer.cpp b/Source/WebCore/platform/graphics/gstreamer/mse/MediaSourcePrivateGStreamer.cpp
index cf0f4d96ba5e..df5abeb29561 100644
--- a/Source/WebCore/platform/graphics/gstreamer/mse/MediaSourcePrivateGStreamer.cpp
+++ b/Source/WebCore/platform/graphics/gstreamer/mse/MediaSourcePrivateGStreamer.cpp
@@ -144,8 +144,22 @@ void MediaSourcePrivateGStreamer::markEndOfStream(EndOfStreamStatus endOfStreamS
 void MediaSourcePrivateGStreamer::unmarkEndOfStream()
 {
     ASSERT(isMainThread());
+    bool didEnd = m_playerPrivate.ended();
     m_playerPrivate.setEosWithNoBuffers(false);
     MediaSourcePrivate::unmarkEndOfStream();
+    if (m_hasAllTracks) {
+        if (didEnd) {
+            GST_DEBUG_OBJECT(m_playerPrivate.pipeline(), "Force flush tracks to clear EOS");
+        } else {
+            GST_WARNING_OBJECT(m_playerPrivate.pipeline(), "Force flush queued samples to clear EOS");
+        }
+        for (auto& privateSourceBuffer : m_sourceBuffers) {
+            auto sourceBuffer = downcast<SourceBufferPrivateGStreamer>(privateSourceBuffer);
+            for (auto& [trackId, _] : sourceBuffer->tracks()) {
+                sourceBuffer->flush(trackId);
+            }
+        }
+    }
 }
 
 MediaPlayer::ReadyState MediaSourcePrivateGStreamer::mediaPlayerReadyState() const
diff --git a/Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp b/Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp
index 90a17aba2a75..67c35a8ebd38 100644
--- a/Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp
+++ b/Source/WebCore/platform/graphics/gstreamer/mse/WebKitMediaSourceGStreamer.cpp
@@ -643,9 +643,8 @@ static void webKitMediaSrcLoop(void* userData)
             GST_DEBUG_OBJECT(pad, "Pushed event was not handled: %" GST_PTR_FORMAT, event.get());
 
         if (isEOS) {
-            DataMutexLocker streamingMembersAfterEOS { stream->streamingMembersDataMutex };
-            GST_INFO_OBJECT(pad, "EOS pushed, resetting wasStreamStartSent to allow re-sending STREAM_START.");
-            streamingMembersAfterEOS->wasStreamStartSent = false;
+            GST_INFO_OBJECT(pad, "EOS pushed, pausing pad task.");
+            gst_pad_pause_task(pad);
         }
     } else
         ASSERT_NOT_REACHED();
-- 
2.53.0

